### ISSUE TITLE
Add support for OR in package.json when clauses

### ISSUE THREAD

**Issue by alexandrudima** [microsoft/vscode#33547]

**Comment by cleidigh** [comment]
@alexandrudima 
Could this be extended for all contexts?  ie Add ContextKeyExpr.Or
This is what I was looking for to make the simpleFind commands work.

**Comment by alexandrudima** [comment]
The keybindings case is a bit more complicated, as we need to compute a reverse lookup. i.e. our UI makes queries of the form: Tell me keybindings that trigger `actions.find`. We then need to consult the keybinding rules for `actions.find` and analyse if there is any other conflict on that specific keybinding. We then need to "prove" if the conflict shadows the keybinding or not. So effectively we need to compute `p -> q`, where `p` is the conflict's `when` clause and `q` is the looked up command's `when` clause. This is typically referred to as an SMT solver, which is quite a complex and costly thing to implement and execute. With limiting the operations inside `p` and `q` to `AND` clauses, it is trivial to implement (that's what we do now).

TL;DR: To investigate how to implement keybindings reverse lookups if a `when` clause contains also `OR` statements.

**Comment by cleidigh** [comment]
@alexandrudima 
Thanks for the great explanation!  I assumed it was no 'accident or oversight' the boolean OR was NOT
included for when classes just have not gone into the resolver to know why yet   - now I know ;-)

**Comment by kevcunnane** [comment]
Support for OR-based clauses is particularly important / useful for tree item filtering in context menus. For example if you have a tree with 5+ viewItem types but want an action on just 2 of them, the below script is much more efficient (support download for files OR folders vs. check for `!root && !item2 && !item3 && ....`).

```json
            "view/item/context": [
                {
                    "command": "ftp.download",
                    "when": "viewItem == file || viewItem == folder",
                    "group": "1tfp@1"
                }]
```

Is it possible to separate out support for OR clauses in Keybindings vs. these cases? If not, is there a recommended workaround such as a callback event on selecting an item that allows us to set a custom context in the code?

**Comment by eamodio** [comment]
@kevcunnane You can currently support this today with the regex support in `when` clauses

```json
     "view/item/context": [
                {
                    "command": "ftp.download",
                    "when": "viewItem =~ /(file|folder)/",
                    "group": "1tfp@1"
                }
     ]
```

You can have any javascript regex pattern in there. You can find examples of all the currently support `when` clauses in my beast of a `package.json` here: https://github.com/eamodio/vscode-gitlens/blob/master/package.json

**Comment by alexandrudima** [comment]
Let's merge with #22778
